lib/CRMS/RightsPredictor.pm

```diff
@@ -0,0 +1,140 @@
+# Returns the following structure:
+#   my $ldc = {
+#     'year' => 1933, # 4-digit int or undef on error
+#     'desc' => 'how derived'
+#   };
+# or undef on error or not enough info.
+sub last_domestic_copyright
+{
+  my $self     = shift;
+  my $year     = shift || ''; # ADD or Pub entered by user
+  my $is_crown = shift; # Crown copyright founds
+
+  if ($year !~ m/^-?\d{1,4}$/) {
+    return { year => undef, desc => "unsupported date format '$year'" };
+  }
+  my $country = $self->{record}->country;
+  if ($country eq 'United Kingdom' || $country eq 'Canada' || $country eq 'Australia') {
+    my $term = 70;
+    if ($is_crown) {
+      $term = 50;
+    } else {
+      if ($country eq 'Canada') {
+        $term = 50 if $year < 1972;
+      } elsif ($country eq 'Australia') {
+        $term = 50 if $year < 1955;
+      }
+    }
+    my $last_domestic_copyright = $year + $term;
+    return { year => $last_domestic_copyright,
+      desc => "Last $country © $last_domestic_copyright ($year + $term-year term)" };
+  }
+  #elsif ($country eq 'Spain') {
+  #  return { year => $year + 80, desc => "$country $year + 80-year term" };
+  #}
+  return { year => undef, desc => "country $country unknown term" };
+}
+
+# CRMS World specific. Predict best radio button (rights combo)
+# to choose based on user selections.
+# Returns data structure with the following fields (all but desc may be absent on error):
+# attr: rights attribute name
+# reason: rights reason name
+# rights: attr/reason string
+# last_domestic_copyright: year
+# desc: last_domestic_copyright desc plus more math, or error message
+sub rights {
+  my $self     = shift;
+  my $year     = shift || ''; # ADD or pub date entered by user
+  my $is_pub   = shift; # Treat $year as pub date for corporate authors
+  my $is_crown = shift; # Crown copyright found
+  #my $pub_year = shift; # Actual pub date when date range, entered by user
+  my $now      = shift || POSIX::strftime("%Y", localtime); # The "current" year
+
+  my $ldc = $self->last_domestic_copyright($year, $is_crown);
+  my $last_domestic_copyright = $ldc->{year};
+  return $ldc unless defined $last_domestic_copyright;
+  my $pub_year;
+  if ($is_pub) {
+    $pub_year = $year;
+  } else {
+    # Verify that we can reason about the year of publication:
+    # If there is a pub date range then it should not span the so-called 1023 boundary
+    $pub_year = $self->{record}->formatPubDate || '';
+    # We currently don't support date ranges at all.
+    # If we collect actual pub dates, we can sanity check that
+    # and then use it in place of the info in the catalog record 008.
+    if ($pub_year !~ m/^-?\d+$/) {
+      return { desc => "unsupported date format '$pub_year'" };
+    }
+  }
+  my ($attr, $reason);
+  my @desc = ( $ldc->{desc} );
+  if ($last_domestic_copyright < $now) {
+    push @desc, "$last_domestic_copyright < $now";
+    my $pub_plus_95 = $pub_year + 95;
```

Comment:
Consider extracting `95` as a named constant -- why 95? Likewise some of the other durations, specific years, etc.

Reply:
Adding
```
use constant {
  # 95 years from date of publication; if published in 1900 the last US copyright year is 1995.
  US_COPYRIGHT_TERM => 95,
  # GATT bestowed full US copyright term to foreign works that were in copyright
  # in the non-US country of publication on Jan 1 1996.
  # Thus, if last source copyright year >= 1996 then the work is GATT-eligible.
  GATT_RESTORATION_START => 1996
};
```